libdnf/hy-subject.c

```diff
@@ -220,73 +220,54 @@ hy_possibilities_next_nevra(HyPossibilities iter, HyNevra *out_nevra)
     return -1;
 }
 
-struct NevraToQuery {
-    int nevra_type;
-    int query_type;
-};
-
-/* Given a subject, attempt to create a query choose the first one, and update
- * the query to try to match it.
- *
- * This code is based on rpm-software-management/dnf/subject.py:get_best_query() at git
- * revision: 1d83fdc0280ca4202281ef489afe600e2f51a32a
- */
 HyQuery
-hy_subject_get_best_query(HySubject subject, DnfSack *sack, gboolean with_provides)
+hy_subject_get_best_solution(HySubject subject, DnfSack *sack, HyForm *forms, HyNevra *nevra,
+                             gboolean icase, gboolean with_nevra, gboolean with_provides,
+                             gboolean with_filenames)
 {
-#if 0
-    int ret = -1;
-    HyNevra nevra = NULL;
-    HyPossibilities iter = NULL;
+    int ret = 0;
     HyQuery query = NULL;
-    DnfReldep *reldep = NULL;
-
-    iter = hy_subject_nevra_possibilities_real(subject, NULL, sack, 0);
-
-    if ((ret = hy_possibilities_next_nevra(iter, &nevra)) < 0)
-        goto out;
-
-    if (nevra != NULL) {
-        struct NevraToQuery nevraquerymap[] =
-            { { HY_NEVRA_NAME, HY_PKG_NAME },
-              { HY_NEVRA_VERSION, HY_PKG_VERSION },
-              { HY_NEVRA_RELEASE, HY_PKG_RELEASE },
-              { HY_NEVRA_ARCH, HY_PKG_ARCH }};
-
-        query = hy_query_create(sack);
 
-        for (guint i = 0; i < G_N_ELEMENTS(nevraquerymap); i++) {
-            const char *str = hy_nevra_get_string(nevra, nevraquerymap[i].nevra_type);
-            if (str)
-                hy_query_filter(query, nevraquerymap[i].query_type, HY_EQ, str);
+    if (with_nevra) {
+        HyPossibilities iter = hy_subject_nevra_possibilities(subject, forms);
+        while (ret != -1) {
+            ret = hy_possibilities_next_nevra(iter, nevra);
+            HyForm form = iter->forms[iter->current];
+            if (ret != -1) {
+                query = hy_nevra_to_query(*nevra, sack, icase);
+                if (hy_query_is_not_empty(query)) {
+                    hy_possibilities_free(iter);
+                    return query;
+                }
+            }
```

Comment:
Probably memory leak. Move line 245 `hy_query_free(query);`  before this line and line 244 is not needed.